src/Streamly/Internal/Unicode/Utf8/Eliminate.hs

```diff
@@ -0,0 +1,414 @@
+findIndex :: (Char -> Bool) -> Utf8 -> Maybe Int
+findIndex p t = unsafePerformIO $ Stream.findIndex p (toStream t)
+
+-- | The 'count' function returns the number of times the
+-- query string appears in the given 'Utf8'. An empty query string is
+-- invalid, and will cause an error to be raised.
+--
+-- In (unlikely) bad cases, this function's time complexity degrades
+-- towards /O(n*m)/.
+--
+-- /Time complexity:/ O(n+m)
+{-# INLINE_NORMAL count #-}
+count :: Utf8 -> Utf8 -> Int
+count = undefined
+
+-- | The 'countChar' function returns the number of times the
+-- query element appears in the given 'Utf8'.
+--
+-- /Time complexity:/ O(n)
+{-# INLINE countChar #-}
+countChar :: Char -> Utf8 -> Int
+countChar c t =
+    unsafePerformIO $ Stream.length $ Stream.filter (== c) (toStream t)
+
+--------------------------------------------------------------------------------
+-- Searching
+--------------------------------------------------------------------------------
+
+--------------------------------------------------------------------------------
+-- Searching with a predicate
+--------------------------------------------------------------------------------
+
+-- | The 'elem' function takes a character and a 'Utf8', and
+-- returns 'True' if the element is found in the given 'Utf8', or
+-- 'False' otherwise.
+--
+-- /Time complexity:/ O(n)
+{-# INLINE elem #-}
+elem :: Char -> Utf8 -> Bool
+elem c = any (== c)
+
+-- | The 'find' function takes a predicate and a 'Utf8', and
+-- returns the first element matching the predicate, or 'Nothing' if
+-- there is no such element.
+--
+-- /Time complexity:/ O(n)
+{-# INLINE find #-}
+find :: (Char -> Bool) -> Utf8 -> Maybe Char
+find p t = unsafePerformIO $ Stream.find p (toStream t)
+
+--------------------------------------------------------------------------------
+-- Predicates
+--------------------------------------------------------------------------------
+
+-- | The 'isPrefixOf' function takes two 'Utf8's and returns
+-- 'True' iff the first is a prefix of the second.
+--
+-- /Time complexity:/ O(n)
+{-# INLINE_NORMAL _isPrefixOf #-}
+_isPrefixOf :: Utf8 -> Utf8 -> Bool
+_isPrefixOf a b =
+    Array.byteLength (toArray a) <= Array.byteLength (toArray b)
+        && unsafePerformIO (Stream.isPrefixOf (toStream a) (toStream b))
+
+-- | The 'isSuffixOf' function takes two 'Utf8's and returns
+-- 'True' iff the first is a suffix of the second.
+--
+-- /Time complexity:/ O(n)
+{-# INLINE _isSuffixOf #-}
+_isSuffixOf :: Utf8 -> Utf8 -> Bool
+_isSuffixOf a b = unsafePerformIO (Stream.isSuffixOf (toStream a) (toStream b))
+
+-- XXX This specific API uses a lot of memory to compile
+-- XXX Use domain specific knowledge to implement it efficiently!
+-- | The 'isInfixOf' function takes two 'Utf8's and returns
+-- 'True' iff the first is contained, wholly and intact, anywhere
+-- within the second.
+--
+-- In (unlikely) bad cases, this function's time complexity degrades
+-- towards /O(n*m)/.
```

Comment:
`Stream.isInfixOf` doc says `O(n)`, why would this be any different? If it is true then we should change that as well.